blockchain/state/iterator.go

```diff
@@ -197,6 +197,137 @@ func (it *NodeIterator) retrieve() bool {
 	return true
 }
 
+// CheckStateConsistencyParallel checks parallelly the consistency of all state/storage trie of given two state database.
+func CheckStateConsistencyParallel(oldDB Database, newDB Database, root common.Hash, quit chan struct{}) error {
+	// Create and iterate a state trie rooted in a sub-node
+	oldState, err := New(root, oldDB)
+	if err != nil {
+		return err
+	}
+
+	_, err = New(root, newDB)
+	if err != nil {
+		return err
+	}
+
+	children, err := oldState.db.TrieDB().NodeChildren(root)
+	if err != nil {
+		logger.Error("CheckStateConsistencyParallel is stop by err", "err", err)
+	}
+
+	logger.Info("CheckStateConsistencyParallel is started", "root", root.String(), "len(children)", len(children))
+
+	resultHashCh := make(chan common.Hash, 10000)
+	resultErrCh := make(chan error)
+
+	for _, child := range children {
+		go concurrentIterator(oldDB, newDB, child, quit, resultHashCh, resultErrCh)
+	}
+
+	var resultErr error
+	cnt := 0
+	logged := time.Now()
+	for finishCnt := 0; finishCnt < len(children); {
+		select {
+		case <-resultHashCh:
+			cnt++
+			if time.Since(logged) > log.StatsReportLimit {
+				logged = time.Now()
+				logger.Info("CheckStateConsistencyParallel progress", "cnt", cnt)
+			}
+		case err := <-resultErrCh:
+			if err != nil {
+				resultErr = err
+				logger.Warn("CheckStateConsistencyParallel got an error", "err", err)
+			}
+
+			finishCnt++
+			logger.Debug("CheckStateConsistencyParallel is being finished", "finishCnt", finishCnt, "err", err)
+		}
+	}
+	logger.Info("CheckStateConsistencyParallel is done", "cnt", cnt, "err", resultErr)
+
+	return resultErr
+}
+
+// concurrentIterator checks the consistency of all state/storage trie of given two state database
+// and pass the result via the channel.
+func concurrentIterator(oldDB Database, newDB Database, root common.Hash, quit chan struct{}, resultCh chan common.Hash, finishCh chan error) (resultErr error) {
+	defer func() {
+		finishCh <- resultErr
+	}()
+
```

Comment:
`CheckStateConsistency` might be called instead of the following code.

Reply:
Right. But we need to modify `CheckStateConsistency`.
I think we can remove `CheckStateConsistency` later. :)